src/Microsoft.PowerShell.Commands.Utility/commands/utility/ConvertFrom-SddlString.cs

```diff
@@ -0,0 +1,187 @@
+
+            PSObject result = new PSObject();
+
+            result.Properties.Add(new PSNoteProperty("Owner", owner));
+            result.Properties.Add(new PSNoteProperty("Group", group));
+            result.Properties.Add(new PSNoteProperty("DiscretionaryAcl", discretionaryAcl));
+            result.Properties.Add(new PSNoteProperty("SystemAcl", systemAcl));
+
+            WriteObject(result);
+        }
+
+        /// <summary>
+        /// Translates a SID into a NT Account
+        /// </summary>
+        private string ConvertToNtAccount(SecurityIdentifier sid)
+        {
+            try
+            {
+                return sid.Translate(typeof(NTAccount)).ToString();
+            }
+            catch {}
+
+            return null;
+        }
+
+        /// <summary>
+        /// Gets the access rights that apply to an access mask, preferring right types
+        /// of 'Type' if specified.
+        /// </summary>
+        private List<string> GetAccessRights(long accessMask, string type)
+        {
+            IDictionary<string, Type> rightTypes = new Dictionary<string, Type>
+            {
+                { "FileSystemRights", typeof(System.Security.AccessControl.FileSystemRights) },
+                { "RegistryRights", typeof(System.Security.AccessControl.RegistryRights) },
+#if !CORECLR
+                { "ActiveDirectoryRights", typeof(System.DirectoryServices.ActiveDirectoryRights) },
+#endif
+                { "MutexRights", typeof(System.Security.AccessControl.MutexRights) },
+                { "SemaphoreRights", typeof(System.Security.AccessControl.SemaphoreRights) },
+#if !CORECLR
+                { "CryptoKeyRights", typeof(System.Security.AccessControl.CryptoKeyRights) },
+#endif
+                { "EventWaitHandleRights", typeof(System.Security.AccessControl.EventWaitHandleRights) }
+            };
+
+            List<Type> typesToExamine = rightTypes.Values.ToList();
+            
+            // If they know the access mask represents a certain type, prefer its names
+            // (i.e.: CreateLink for the registry over CreateDirectories for the filesystem)
+            if (!String.IsNullOrEmpty(type))
+            {
+                typesToExamine.Insert(0, rightTypes[type]);
+            }
+            
+            // Stores the access types we've found that apply
+            List<string> foundAccess = new List<string>();
+
+            // Store the access types we've already seen, so that we don't report access
+            // flags that are essentially duplicate. Many of the access values in the different
+            // enumerations have the same value but with different names.
+            HashSet<long> foundValues = new HashSet<long>();
+
+            // Go through the entries in the different right types, and see if they apply to the
+            // provided access mask. If they do, then add that to the result.
+            foreach (var rightType in typesToExamine)
```

Comment:
@iSazonov Apart from fixing CommonSecurityDescriptor constructor call (by requiring isContainer and isDS flags) we should also take care about correct parsing of AccessMasks to enum names. FileSystemRights and ActiveDirectoryRights contain several names that use the same value, eg. FileSystemRights contain AppendData & CreateDirectories which have [the same value (4)](https://github.com/dotnet/corefx/blob/master/src/System.IO.FileSystem.AccessControl/ref/System.IO.FileSystem.AccessControl.cs#L83&L85). In consequence Directories with access flag (4) can get AppendData right, what is incorrect.

As a solution we can add custom enums (or at least custom mapping dictionaries) containing distinct subsets of allowed values - one for containers (ContainerSystemRights) and one for other objects (FileSystemRights). Alternatively we can discard CommonSecurityDescriptor/Right-Enum based parsing algorithm entirely and try to implement text parser using sddl docs. What do you think ?